src/services/stringCompletions.ts

```diff
@@ -1182,8 +1182,6 @@ function getCompletionEntriesFromTypings(host: LanguageServiceHost, options: Com
 }
 
 function enumerateNodeModulesVisibleToScript(host: LanguageServiceHost, scriptPath: string): readonly string[] {
-    if (!host.readFile || !host.fileExists) return emptyArray;

```

Comment:
The LanguageServiceHost has a comment

```
     * Unlike `realpath and `readDirectory`, `readFile` and `fileExists` are now _required_
     * to properly acquire and setup source files under module: node16+ modes.
```

so I believe these checks have been unnecessary since https://github.com/microsoft/TypeScript/commit/94724a8c2e68a4c7e267072ca79971f317c45e4a#diff-9717eb6d935579eea3b5662088dd258d4e1e17e2a405c3f1fa431e981e01b08f

Reply:
I would git blame this; node16 is a "recent" addition and requiring those would be new, so this is likely a compatibility shim to ensure that older callers don't crash. We do this sort of thing somewhat often when making "breaking" API changes (but, still detecting older uses and trying to fix them within reason).

Reply:
It looks like it went in 6 years ago with https://github.com/microsoft/TypeScript/commit/70682b77991d02a3ba9b2438397d6bd82cf5637f#diff-10538e33f93edc9b510a4cd505ecaa718f6ab878a54b7d044a18ef10b1e36792R383 . 

Better understanding the compatibility shim, I'll undo the changes that assume `host.readFile` and `host.fileExists`. I wouldn't want an innocent change like this to break someone's module compatibility.